### ISSUE TITLE
extra_context stopped working for me in 1.2.1

### ISSUE THREAD

**Issue by xsteadfastx** [audreyr/cookiecutter#571]
Im sure im missing a change but i cant find what its wrong with this snippet im using:

```
from cookiecutter.main import cookiecutter
from datetime import datetime
from tzlocal import get_localzone
import pytz


cookiecutter(
    'cookiecutter-new-post',
    overwrite_if_exists=True,
    extra_context={
        'date': datetime.utcnow().replace(tzinfo=pytz.utc).astimezone(get_localzone()).strftime('%Y-%m-%d %H:%M')})
```
It doesnt create `{{ cookiecutter.date }}`. This works fine with version 1.1.0. Maybe someone has a hint for me. And by the way: thanks for this beautiful piece of software. I use it everyday and it makes everything so much easier :smile:

**Comment by hackebrot** [comment]
Hi @xsteadfastx! Thank you for your feedback :+1:

Happy to hear that you like **Cookiecutter** :smile:

You should be fine by adding ``date`` to your ``cookiecutter.json``. I *broke* this *feature* in 8f08298 in the course of #542. The rationale of this change is that template authors should declare all the used variable in the context file... :no_mouth: 

HTH

**Comment by xsteadfastx** [comment]
this did the trick :) thanks alot

**Comment by hackebrot** [comment]
You're welcome! :smiley_cat:

**Comment by danielsamuels** [comment]
Hey,

I'm just wondering what the rationale was behind not adding values to the context if they're not defined in the JSON file? 

We're calling `cookiecutter()` using the API and passing in a dict of values that we need in our hooks & templates, but now we have to skip past a whole bunch of values in the setup script, even if we know that they will never be defined during initialisation, nor will they ever be changed.  

An example of how this applies to us is that we generate a Github project via the API before we create the project, we then pass the clone URL into the extra context so we can use it within a post_gen hook to set the generated project's remote. This is a value that clearly will never be user-defined, but is still 100% necessary to be passed into the Jinja2 context, but now doesn't work due to the aforementioned change.

Is there any way to get this working again without having to downgrade our Cookiecutter version & without making it a key in the cookiecutter.json file?

Thanks,
Daniel.

**Comment by pydanny** [comment]
@danielsamuels My deepest regrets, but I'm unable to wrap my head around the double negatives in the first sentence. Since the rest of your question seems to build on it, I want to make certain I understand it before answering. Could you rephrase your question?

**Comment by danielsamuels** [comment]
Hi @pydanny,

My question follows on from the first two comments in this pull request, but I'll see if I can make things clearer with some practical examples.

Consider a cookiecutter.json file which looks like this:

```json
{
    "repo_name": "foo"
}
```

In versions of Cookiecutter before 8f08298, you were able to call the Cookiecutter API as follows:

```python
cookiecutter(
    'gh:user/repo',
    extra_context={
        "foo": "bar"
    }
)
```

..and that would allow you to use `{{ cookiecutter.foo }}` within your project files, to have it output "bar" as you might expect. However, the change made in 8f08298 means that `foo` is no longer available within the Jinja2 context because it's not listed as a field in the cookiecutter.json.  

Going back to my comment, I'm calling a whole bunch of different APIs, or using Selenium to configure various third-party services and then passing the output of those APIs into the `extra_context` of my `cookiecutter` call. For clarity, I have two cookiecutter repos, one which does the API work (because I can't add to the Jinja2 context from within a pre_gen hook) and the other which contains the actual project files. The first project generates the second. This code is all public, so you can see it in action [here](https://github.com/onespacemedia/developer-automation/blob/2d6abdc4728221d29ccac12834bb86f5ac47517a/hooks/post_gen_project.py) (though the extra_context stuff isn't in this particular version of the file, you can see how it will work in [this version](https://github.com/onespacemedia/developer-automation/blob/7ce078d6db337dc5452e4a65b4aec841fd03dd09/hooks/post_gen_project.py)).

I hope this makes things clearer, let me know if you have any further questions.

Thanks,
Daniel.

**Comment by hackebrot** [comment]
Hi @danielsamuels! :wave: 

Please see PR #597 for an explanation on 8f082981fdcb3d1138d2962de88622006070ae91 and also a change to bring back the functionality for extra context.
